apiserver/httpcontext.go

```diff
@@ -111,9 +111,65 @@ func checkPermissions(tag names.Tag, acceptFunc common.GetAuthFunc) (bool, error
 	return false, errors.NotValidf("tag kind %v", tag.Kind())
 }
 
-// stateForRequestAuthenticatedUser is like stateForRequestAuthenticated
+// stateForMigration asserts that the incoming connection is from a user that
+// has admin permissions on the controller model. The method also gets the
+// model uuid for the model being migrated from a request header, and returns
+// the state instance for that model.
+func (ctxt *httpContext) stateForMigration(r *http.Request) (st *state.State, err error) {
+	var user state.Entity
+	st, user, err = ctxt.stateAndEntityForRequestAuthenticatedUser(r)
+	if err != nil {
+		return nil, err
+	}
+	defer func() {
+		if err != nil {
+			ctxt.release(st)
```

Comment:
I think this will hit the same problem I did - because you're using the return variable to store the *State in, it gets overwritten by the return before the defer fires. If you run the tests, any tests of error cases will have nil pointer derefs. Changing the name of the return value should fix it.